src/app/heroes/heroes.component.spec.ts

```diff
@@ -0,0 +1,62 @@
+import { HeroesComponent } from './heroes.component';
+import { Hero } from '../hero';
+import { HeroService } from '../hero.service';
+import { of } from 'rxjs';
+
+describe('Heroes Component', ()=> {
+
+  let HEROES : Hero[] = [
+    {id: 1, name: 'Spiderman', strength: 15},
+    {id: 2, name: 'Batman', strength: 10},
+    {id: 3, name: 'Deadpool', strength: 3},
+  ];
+
+  // let mockHeroSvc : HeroService;
+  // let heroesCmp: HeroesComponent;
+  /*
+    * work with above his test is work
+    * in the last lesson became as below
+    * No way pass with me
+  */
+
+  let mockHeroSvc ;
+  let heroesCmp ;
+
+  beforeEach(()=> {
+    // #question how to fix this error
+    // mockHeroSvc = {
+    //   deleteHero: function() {
+    //     return { subscribe: function() {}}
+    //   }
+    // };
+
+    mockHeroSvc = jasmine.createSpyObj<HeroService>(['deleteHero']);
+    mockHeroSvc.deleteHero.and.returnValue(of(true));
+
+    heroesCmp = new HeroesComponent(mockHeroSvc);
+
+  })
+
+  it('should remove the selected hero from the heros list', ()=> {
+    // arrang    
+    heroesCmp.heroes = HEROES;
+
+    // act
+    heroesCmp.delete(HEROES[1]);
+
+    // assert
+    expect(heroesCmp.heroes.length).toBe(2);
+    expect(heroesCmp.heroes[0]).toBe(HEROES[0]);
+    expect(heroesCmp.heroes[1]).toBe(HEROES[2]);
+  })
+
+  it('should call delelteHero with the correct hero', ()=> {
```

Comment:
Remove this test from here. You should test one unit in each test. here, i guess your main concern is testing the component class with its template. And since the class has an injected service, you've provided a mocked service using jasmine's spy objects.

![image](https://user-images.githubusercontent.com/35868173/110379106-0e97a600-805f-11eb-8ddc-e2d9ffa5dd66.png)

When you try to test the service methods, you get the error since the type of the mocked service is any and you're trying to access `deleteHero`  method from it which will definitely throw error. And if you explicitly give the mocked service type of `HeroService`, you wont be able to access`.and.returnValues` in the spyObject you just created since the `HeroService` does not have a property with this name.

![image](https://user-images.githubusercontent.com/35868173/110379739-df356900-805f-11eb-81fb-9e0741461938.png)


All in all, you should make a separate unit test for this service if you want to test its method.
Here is a link in angular doc for doing so: https://angular.io/guide/testing-services